{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: genoml2 in c:\\users\\cnrm\\documents\\github\\biof309\\genoml2 (1.0.0b4)\n",
      "Requirement already satisfied: joblib in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from genoml2) (0.13.2)\n",
      "Requirement already satisfied: matplotlib in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from genoml2) (3.1.1)\n",
      "Requirement already satisfied: numpy in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from genoml2) (1.16.5)\n",
      "Requirement already satisfied: tables in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from genoml2) (3.5.2)\n",
      "Requirement already satisfied: pandas in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from genoml2) (0.25.1)\n",
      "Requirement already satisfied: pandas_plink in c:\\users\\cnrm\\appdata\\roaming\\python\\python37\\site-packages (from genoml2) (2.2.1)\n",
      "Requirement already satisfied: requests in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from genoml2) (2.22.0)\n",
      "Requirement already satisfied: scikit-learn in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from genoml2) (0.21.3)\n",
      "Requirement already satisfied: scipy in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from genoml2) (1.3.1)\n",
      "Requirement already satisfied: seaborn in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from genoml2) (0.9.0)\n",
      "Requirement already satisfied: statsmodels in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from genoml2) (0.10.1)\n",
      "Requirement already satisfied: xgboost in c:\\users\\cnrm\\appdata\\roaming\\python\\python37\\site-packages (from genoml2) (1.2.1)\n",
      "Requirement already satisfied: cycler>=0.10 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from matplotlib->genoml2) (0.10.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from matplotlib->genoml2) (1.1.0)\n",
      "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from matplotlib->genoml2) (2.4.2)\n",
      "Requirement already satisfied: python-dateutil>=2.1 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from matplotlib->genoml2) (2.8.0)\n",
      "Requirement already satisfied: numexpr>=2.6.2 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from tables->genoml2) (2.7.0)\n",
      "Requirement already satisfied: six>=1.9.0 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from tables->genoml2) (1.12.0)\n",
      "Requirement already satisfied: mock>=2.0 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from tables->genoml2) (3.0.5)\n",
      "Requirement already satisfied: pytz>=2017.2 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from pandas->genoml2) (2019.3)\n",
      "Requirement already satisfied: zstandard>=0.13.0 in c:\\users\\cnrm\\appdata\\roaming\\python\\python37\\site-packages (from pandas_plink->genoml2) (0.14.0)\n",
      "Requirement already satisfied: tqdm>=4.36.1 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from pandas_plink->genoml2) (4.36.1)\n",
      "Requirement already satisfied: pytest>=5.2.2 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from pandas_plink->genoml2) (6.1.1)\n",
      "Requirement already satisfied: cffi>=1.14.3 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from pandas_plink->genoml2) (1.14.3)\n",
      "Requirement already satisfied: dask[array,dataframe]>=2.6.0 in c:\\users\\cnrm\\appdata\\roaming\\python\\python37\\site-packages (from pandas_plink->genoml2) (2.30.0)\n",
      "Requirement already satisfied: xarray>=0.16.1 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from pandas_plink->genoml2) (0.16.1)\n",
      "Requirement already satisfied: Deprecated>=1.2.6 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from pandas_plink->genoml2) (1.2.10)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from requests->genoml2) (2019.9.11)\n",
      "Requirement already satisfied: idna<2.9,>=2.5 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from requests->genoml2) (2.8)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from requests->genoml2) (1.24.2)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from requests->genoml2) (3.0.4)\n",
      "Requirement already satisfied: patsy>=0.4.0 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from statsmodels->genoml2) (0.5.1)\n",
      "Requirement already satisfied: setuptools in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from kiwisolver>=1.0.1->matplotlib->genoml2) (41.4.0)\n",
      "Requirement already satisfied: packaging in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from pytest>=5.2.2->pandas_plink->genoml2) (19.2)\n",
      "Requirement already satisfied: attrs>=17.4.0 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from pytest>=5.2.2->pandas_plink->genoml2) (19.2.0)\n",
      "Requirement already satisfied: iniconfig in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from pytest>=5.2.2->pandas_plink->genoml2) (1.1.1)\n",
      "Requirement already satisfied: importlib-metadata>=0.12; python_version < \"3.8\" in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from pytest>=5.2.2->pandas_plink->genoml2) (0.23)\n",
      "Requirement already satisfied: atomicwrites>=1.0; sys_platform == \"win32\" in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from pytest>=5.2.2->pandas_plink->genoml2) (1.3.0)\n",
      "Requirement already satisfied: pluggy<1.0,>=0.12 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from pytest>=5.2.2->pandas_plink->genoml2) (0.13.0)\n",
      "Requirement already satisfied: toml in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from pytest>=5.2.2->pandas_plink->genoml2) (0.10.1)\n",
      "Requirement already satisfied: py>=1.8.2 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from pytest>=5.2.2->pandas_plink->genoml2) (1.9.0)\n",
      "Requirement already satisfied: colorama; sys_platform == \"win32\" in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from pytest>=5.2.2->pandas_plink->genoml2) (0.4.1)\n",
      "Requirement already satisfied: pycparser in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from cffi>=1.14.3->pandas_plink->genoml2) (2.19)\n",
      "Requirement already satisfied: pyyaml in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from dask[array,dataframe]>=2.6.0->pandas_plink->genoml2) (5.1.2)\n",
      "Requirement already satisfied: toolz>=0.8.2; extra == \"array\" in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from dask[array,dataframe]>=2.6.0->pandas_plink->genoml2) (0.10.0)\n",
      "Requirement already satisfied: partd>=0.3.10; extra == \"dataframe\" in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from dask[array,dataframe]>=2.6.0->pandas_plink->genoml2) (1.0.0)\n",
      "Requirement already satisfied: fsspec>=0.6.0; extra == \"dataframe\" in c:\\users\\cnrm\\appdata\\roaming\\python\\python37\\site-packages (from dask[array,dataframe]>=2.6.0->pandas_plink->genoml2) (0.8.4)\n",
      "Requirement already satisfied: wrapt<2,>=1.10 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from Deprecated>=1.2.6->pandas_plink->genoml2) (1.11.2)\n",
      "Requirement already satisfied: zipp>=0.5 in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from importlib-metadata>=0.12; python_version < \"3.8\"->pytest>=5.2.2->pandas_plink->genoml2) (0.6.0)\n",
      "Requirement already satisfied: locket in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from partd>=0.3.10; extra == \"dataframe\"->dask[array,dataframe]>=2.6.0->pandas_plink->genoml2) (0.2.0)\n",
      "Requirement already satisfied: more-itertools in c:\\users\\cnrm\\anaconda3\\lib\\site-packages (from zipp>=0.5->importlib-metadata>=0.12; python_version < \"3.8\"->pytest>=5.2.2->pandas_plink->genoml2) (7.2.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install genoml2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-10-503b9b6b3c75>, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-10-503b9b6b3c75>\"\u001b[1;36m, line \u001b[1;32m1\u001b[0m\n\u001b[1;33m    genoml discrete supervised munge \\\u001b[0m\n\u001b[1;37m                  ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "genoml discrete supervised munge \\\n",
    "--prefix ./outputs/test_discrete_geno \\\n",
    "--geno ./examples/discrete/training \\\n",
    "--pheno ./examples/discrete/training_pheno.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "from genoml import preprocessing\n",
    "from genoml import discrete\n",
    "from genoml import continuous\n",
    "from genoml import cli\n",
    "\n",
    "import argparse\n",
    "import functools\n",
    "import sys\n",
    "\n",
    "from genoml import utils\n",
    "from genoml import dependencies\n",
    "from genoml.cli import continuous_supervised_train, continuous_supervised_tune, \\\n",
    "    continuous_supervised_test, discrete_supervised_train, \\\n",
    "    discrete_supervised_tune, discrete_supervised_test, munging, harmonizing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copyright 2020 The GenoML Authors. All Rights Reserved.\n",
    "#\n",
    "# Licensed under the Apache License, Version 2.0 (the \"License\");\n",
    "# you may not use this file except in compliance with the License.\n",
    "# You may obtain a copy of the License at\n",
    "#\n",
    "#     http://www.apache.org/licenses/LICENSE-2.0\n",
    "#\n",
    "# Unless required by applicable law or agreed to in writing, software\n",
    "# distributed under the License is distributed on an \"AS IS\" BASIS,\n",
    "# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
    "# See the License for the specific language governing permissions and\n",
    "# limitations under the License.\n",
    "# ==============================================================================\n",
    "\n",
    "\n",
    "import argparse\n",
    "import functools\n",
    "import sys\n",
    "\n",
    "from genoml import utils\n",
    "from genoml import dependencies\n",
    "from genoml.cli import continuous_supervised_train, continuous_supervised_tune, \\\n",
    "    continuous_supervised_test, discrete_supervised_train, \\\n",
    "    discrete_supervised_tune, discrete_supervised_test, munging, harmonizing\n",
    "\n",
    "\n",
    "def handle_main():\n",
    "    entry_points = [\n",
    "        {\"name\": \"continuous\", \"handler\": handle_continuous,\n",
    "        \"description\": \"for processing continuous datatypes (ex: age at onset)\"},\n",
    "        {\"name\": \"discrete\", \"handler\": handle_discrete, \n",
    "        \"description\": \"for processing discrete datatypes (ex: case vs. control status)\"},\n",
    "        {\"name\": \"harmonize\", \"handler\": handle_harmonize,\n",
    "        \"description\": \"for harmonizing incoming test datasets to use the same SNPs and reference alleles prior to munging, training, and testing\"},\n",
    "    ]\n",
    "    handle_dispatcher(entry_points, \"genoml\", 0)\n",
    "\n",
    "\n",
    "def handle_continuous():\n",
    "    entry_points = [\n",
    "        {\"name\": \"supervised\", \"handler\": handle_continuous_supervised,\n",
    "         \"description\": \"choice of munge, train, tune, or test your continuous dataset must be specified\"},\n",
    "    ]\n",
    "    handle_dispatcher(entry_points, \"genoml continuous\", 1)\n",
    "\n",
    "\n",
    "def handle_discrete():\n",
    "    entry_points = [\n",
    "        {\"name\": \"supervised\", \"handler\": handle_discrete_supervised,\n",
    "         \"description\": \"choice of munge, train, tune, or test your discrete dataset must be specified\"},\n",
    "    ]\n",
    "    handle_dispatcher(entry_points, \"genoml discrete\", 1)\n",
    "\n",
    "\n",
    "def handle_continuous_supervised():\n",
    "    entry_points = [\n",
    "        {\"name\": \"munge\", \"handler\": handle_continuous_supervised_munge,\n",
    "         \"description\": \"for supervised munging and preprocessing of your continuous dataset prior to training\"},\n",
    "        {\"name\": \"train\", \"handler\": handle_continuous_supervised_train,\n",
    "         \"description\": \"for supervised training of your munged continuous dataset by competing different algorithms\"},\n",
    "        {\"name\": \"tune\", \"handler\": handle_continuous_supervised_tune,\n",
    "         \"description\": \"for supervised tuning of your munged and trained continuous dataset\"},\n",
    "        {\"name\": \"test\", \"handler\": handle_continuous_supervised_test,\n",
    "         \"description\": \"for supervised testing of your model generated on continuous data on unseen continuous data (after its been harmonized and munged)\"},\n",
    "    ]\n",
    "    handle_dispatcher(entry_points, \"genoml continuous supervised\", 2)\n",
    "\n",
    "\n",
    "def handle_discrete_supervised():\n",
    "    entry_points = [\n",
    "        {\"name\": \"munge\", \"handler\": handle_discrete_supervised_munge,\n",
    "         \"description\": \"for supervised munging and preprocessing of your discrete dataset prior to training\"},\n",
    "        {\"name\": \"train\", \"handler\": handle_discrete_supervised_train,\n",
    "         \"description\": \"for supervised training of your munged discrete dataset by competing different algorithms\"},\n",
    "        {\"name\": \"tune\", \"handler\": handle_discrete_supervised_tune,\n",
    "         \"description\": \"for supervised tuning of your munged and trained discrete dataset\"},\n",
    "        {\"name\": \"test\", \"handler\": handle_discrete_supervised_test,\n",
    "         \"description\": \"for supervised testing of your model generated on discrete data on unseen discrete data (after its been harmonized and munged)\"},\n",
    "    ]\n",
    "    handle_dispatcher(entry_points, \"genoml discrete supervised\", 2)\n",
    "\n",
    "\n",
    "def handle_harmonize():\n",
    "    handle_endpoints(\"genoml harmonize\",\n",
    "                     [\"test_geno_prefix\", \"test_prefix\", \"ref_model_prefix\",\n",
    "                      \"training_snps_alleles\"],\n",
    "                     harmonizing.main, 1)\n",
    "\n",
    "\n",
    "def handle_continuous_supervised_munge():\n",
    "    handle_endpoints(\"genoml continuous supervised munge\",\n",
    "                     [\"prefix\", \"impute\", \"geno\", \"skip_prune\", \"pheno\", \"addit\",\n",
    "                      \"feature_selection\", \"gwas\", \"p\", \"vif\", \"iter\",\n",
    "                      \"ref_cols_harmonize\"],\n",
    "                     functools.partial(munging.main, data_type=\"c\"), 3)\n",
    "\n",
    "\n",
    "def handle_continuous_supervised_train():\n",
    "    handle_endpoints(\"genoml continuous supervised train\",\n",
    "                     [\"prefix\", \"export_predictions\", \"matching_columns\"],\n",
    "                     continuous_supervised_train.main, 3)\n",
    "\n",
    "\n",
    "def handle_continuous_supervised_tune():\n",
    "    handle_endpoints(\"genoml continuous supervised tune\",\n",
    "                     [\"prefix\", \"max_tune\", \"n_cv\"],\n",
    "                     continuous_supervised_tune.main, 3)\n",
    "\n",
    "\n",
    "def handle_continuous_supervised_test():\n",
    "    handle_endpoints(\"genoml continuous supervised test\",\n",
    "                     [\"prefix\", \"test_prefix\", \"ref_model_prefix\"],\n",
    "                     continuous_supervised_test.main, 3)\n",
    "\n",
    "\n",
    "def handle_discrete_supervised_munge():\n",
    "    handle_endpoints(\"genoml discrete supervised munge\",\n",
    "                     [\"prefix\", \"impute\", \"geno\", \"skip_prune\", \"pheno\", \"addit\",\n",
    "                      \"feature_selection\", \"gwas\", \"p\", \"vif\", \"iter\",\n",
    "                      \"ref_cols_harmonize\"],\n",
    "                     functools.partial(munging.main, data_type=\"d\"), 3)\n",
    "\n",
    "\n",
    "def handle_discrete_supervised_train():\n",
    "    handle_endpoints(\"genoml discrete supervised train\",\n",
    "                     [\"prefix\", \"metric_max\", \"prob_hist\", \"auc\",\n",
    "                      \"matching_columns\"],\n",
    "                     discrete_supervised_train.main, 3)\n",
    "\n",
    "\n",
    "def handle_discrete_supervised_tune():\n",
    "    handle_endpoints(\"genoml discrete supervised tune\",\n",
    "                     [\"prefix\", \"metric_tune\", \"max_tune\", \"n_cv\"],\n",
    "                     discrete_supervised_tune.main, 3)\n",
    "\n",
    "\n",
    "def handle_discrete_supervised_test():\n",
    "    handle_endpoints(\"genoml discrete supervised test\",\n",
    "                     [\"prefix\", \"test_prefix\", \"ref_model_prefix\"],\n",
    "                     discrete_supervised_test.main, 3)\n",
    "\n",
    "\n",
    "def handle_dispatcher(entry_points, command_name, level):\n",
    "    usage_description = f'{command_name} <command> [<args>]\\n'\n",
    "    for command in entry_points:\n",
    "        usage_description += \"   {name:15s} {description}\\n\".format(**command)\n",
    "\n",
    "    parser = argparse.ArgumentParser(prog=command_name,\n",
    "                                     description=f'{command_name}',\n",
    "                                     usage=usage_description\n",
    "                                     )\n",
    "\n",
    "    parser.add_argument('command', help='Subcommand to run')\n",
    "    args = parser.parse_args(sys.argv[level + 1:level + 2])\n",
    "\n",
    "    candidates = []\n",
    "    for command in entry_points:\n",
    "        if command[\"name\"] == args.command:\n",
    "            command[\"handler\"]()\n",
    "            return\n",
    "        if command[\"name\"].startswith(args.command):\n",
    "            candidates.append(command)\n",
    "\n",
    "    if len(candidates) == 1:\n",
    "        candidates[0][\"handler\"]()\n",
    "        return\n",
    "\n",
    "    parser.print_usage()\n",
    "    print(f'Unrecognized command: {args.command}')\n",
    "    exit(1)\n",
    "\n",
    "\n",
    "def add_default_flag(parser, flag_name):\n",
    "    if flag_name == \"prefix\":\n",
    "        parser.add_argument('--prefix', type=str, default=\"GenoML_data\",\n",
    "                            help=\"Prefix for your output build.\")\n",
    "\n",
    "    elif flag_name == \"metric_max\":\n",
    "        parser.add_argument('--metric_max', type=str, default='AUC',\n",
    "                            choices=['AUC', \"Balanced_Accuracy\", \"Specificity\",\n",
    "                                     \"Sensitivity\"],\n",
    "                            help='How do you want to determine which algorithm'\n",
    "                                 ' performed the best? [default: AUC].')\n",
    "\n",
    "    elif flag_name == \"verbose\":\n",
    "        parser.add_argument('-v', '--verbose', action='store_true',\n",
    "                            default=False, help=\"Verbose output.\")\n",
    "\n",
    "    elif flag_name == \"matching_columns\":\n",
    "        parser.add_argument('--matching_columns', type=str, default=None,\n",
    "                            help=\"This is the list of intersecting columns \"\n",
    "                                 \"between reference and testing datasets with \"\n",
    "                                 \"the suffix *_finalHarmonizedCols_toKeep.txt\")\n",
    "\n",
    "    elif flag_name == \"prob_hist\":\n",
    "        parser.add_argument('--prob_hist', type=bool, default=False)\n",
    "\n",
    "    elif flag_name == \"auc\":\n",
    "        parser.add_argument('--auc', type=bool, default=False)\n",
    "\n",
    "    elif flag_name == \"export_predictions\":\n",
    "        parser.add_argument('--export_predictions', type=bool, default=False)\n",
    "\n",
    "    elif flag_name == \"metric_tune\":\n",
    "        parser.add_argument('--metric_tune', type=str, default='AUC',\n",
    "                            choices=['AUC', \"Balanced_Accuracy\"],\n",
    "                            help='Using what metric of the best algorithm do '\n",
    "                                 'you want to tune on? [default: AUC].')\n",
    "\n",
    "    elif flag_name == \"max_tune\":\n",
    "        parser.add_argument('--max_tune', type=int, default=50,\n",
    "                            help='Max number of tuning iterations: (integer '\n",
    "                                 'likely greater than 10). This governs the '\n",
    "                                 'length of tuning process, run speed and the '\n",
    "                                 'maximum number of possible combinations of '\n",
    "                                 'tuning parameters [default: 50].')\n",
    "\n",
    "    elif flag_name == \"n_cv\":\n",
    "        parser.add_argument('--n_cv', type=int, default=5,\n",
    "                            help='Number of cross validations: (integer likely '\n",
    "                                 'greater than 3). Here we set the number of '\n",
    "                                 'cross-validation runs for the algorithms '\n",
    "                                 '[default: 5].')\n",
    "\n",
    "    elif flag_name == \"test_prefix\":\n",
    "        parser.add_argument('--test_prefix', type=str, default='GenoML_data',\n",
    "                            help='Prefix for the dataset you would like to '\n",
    "                                 'test against your reference model. '\n",
    "                                 'Remember, the model will not function well '\n",
    "                                 'if it does not include the same features, '\n",
    "                                 'and these features should be on the same '\n",
    "                                 'numeric scale, you can leave off the '\n",
    "                                 '\\'.dataForML.h5\\' suffix.')\n",
    "\n",
    "    elif flag_name == \"ref_model_prefix\":\n",
    "        parser.add_argument('--ref_model_prefix', type=str,\n",
    "                            default='GenoML_model',\n",
    "                            help='Prefix of your reference model file, '\n",
    "                                 'you can leave off the \\'.joblib\\' suffix.')\n",
    "\n",
    "    elif flag_name == \"test_geno_prefix\":\n",
    "        parser.add_argument('--test_geno_prefix', type=str,\n",
    "                            default='genotype_binaries',\n",
    "                            help='Prefix of the genotypes for the test '\n",
    "                                 'dataset in PLINK binary format.',\n",
    "                            required=True)\n",
    "\n",
    "    elif flag_name == \"training_snps_alleles\":\n",
    "        parser.add_argument('--training_snps_alleles', type=str,\n",
    "                            default=None,\n",
    "                            help='File to the SNPs and alleles file generated '\n",
    "                                 'in the training phase that we will use to '\n",
    "                                 'compare.',\n",
    "                            required=True)\n",
    "\n",
    "    elif flag_name == \"pheno\":\n",
    "        parser.add_argument(\"--pheno\", type=str, default=\"lost\",\n",
    "                            help=\"Phenotype: (string file path). Path to CSV \"\n",
    "                                 \"phenotype file [default: lost].\",\n",
    "                            required=True)\n",
    "\n",
    "    elif flag_name == \"geno\":\n",
    "        parser.add_argument(\"--geno\", type=str, default=None,\n",
    "                            help=\"Genotype: (string file path). Path to PLINK \"\n",
    "                                 \"format genotype file, everything before the \"\n",
    "                                 \"*.bed/bim/fam [default: None].\")\n",
    "\n",
    "    elif flag_name == \"skip_prune\":\n",
    "        parser.add_argument(\"--skip_prune\", type=str, default=\"no\",\n",
    "                            help=\"[default: no].\",\n",
    "                            required=False)\n",
    "                            \n",
    "    elif flag_name == \"addit\":\n",
    "        parser.add_argument(\"--addit\", type=str, default=None,\n",
    "                            help=\"Additional: (string file path). Path to CSV \"\n",
    "                                 \"format feature file [default: None].\")\n",
    "\n",
    "    elif flag_name == \"gwas\":\n",
    "        parser.add_argument(\"--gwas\", type=str, default=None,\n",
    "                            help=\"GWAS summary stats: (string file path). \"\n",
    "                                 \"Path to CSV format external GWAS summary \"\n",
    "                                 \"statistics containing at least the columns \"\n",
    "                                 \"SNP and P in the header [default: nope].\")\n",
    "\n",
    "    elif flag_name == \"p\":\n",
    "        parser.add_argument(\"--p\", type=float, default=0.001,\n",
    "                            help=\"P threshold for GWAS: (some value between \"\n",
    "                                 \"0-1). P value to filter your SNP data on [\"\n",
    "                                 \"default: 0.001].\")\n",
    "\n",
    "    elif flag_name == \"vif\":\n",
    "        parser.add_argument(\"--vif\", type=int, default=0,\n",
    "                            help=\"Variance Inflation Factor (VIF): (integer). \"\n",
    "                                 \"This is the VIF threshold for pruning \"\n",
    "                                 \"non-genotype features. We recommend a value \"\n",
    "                                 \"of 5-10. The default of 0 means no VIF \"\n",
    "                                 \"filtering will be done. [default: 0].\")\n",
    "\n",
    "    elif flag_name == \"iter\":\n",
    "        parser.add_argument(\"--iter\", type=int, default=0,\n",
    "                            help=\"Iterator: (integer). How many iterations of \"\n",
    "                                 \"VIF pruning of features do you want to run. \"\n",
    "                                 \"To save time VIF is run in randomly \"\n",
    "                                 \"assorted chunks of 1000 features per \"\n",
    "                                 \"iteration. The default of 1 means only one \"\n",
    "                                 \"pass through the data. [default: 1].\")\n",
    "\n",
    "    elif flag_name == \"impute\":\n",
    "        parser.add_argument(\"--impute\", type=str, default=\"median\",\n",
    "                            help=\"Imputation: (mean, median). Governs \"\n",
    "                                 \"secondary imputation and data \"\n",
    "                                 \"transformation [default: median].\",\n",
    "                            choices=[\"median\", \"mean\"])\n",
    "\n",
    "    elif flag_name == \"feature_selection\":\n",
    "        parser.add_argument('--feature_selection', type=int, default=0,\n",
    "                            help='Run a quick tree-based feature selection '\n",
    "                                 'routine prior to anything else, here you '\n",
    "                                 'input the integer number of estimators '\n",
    "                                 'needed, we suggest >= 50. The default of 0 '\n",
    "                                 'will skip this functionality. This will '\n",
    "                                 'also output a reduced dataset for analyses '\n",
    "                                 'in addition to feature ranks. [default: 0]')\n",
    "\n",
    "    elif flag_name == \"ref_cols_harmonize\":\n",
    "        parser.add_argument('--ref_cols_harmonize', type=str, default=None,\n",
    "                            help='Are you now munging a test dataset '\n",
    "                                 'following the harmonize step? Here you '\n",
    "                                 'input the path to the to the '\n",
    "                                 '*_refColsHarmonize_toKeep.txt file '\n",
    "                                 'generated at that step.')\n",
    "\n",
    "    else:\n",
    "        raise Exception(f\"Unknown flag: {flag_name}\")\n",
    "\n",
    "\n",
    "def handle_endpoints(command_name, flag_names, endpoint, level):\n",
    "    parser = argparse.ArgumentParser(prog=command_name)\n",
    "    for name in flag_names:\n",
    "        add_default_flag(parser, name)\n",
    "\n",
    "    add_default_flag(parser, \"verbose\")\n",
    "\n",
    "    args, unknown = parser.parse_known_args(sys.argv[level + 1:])\n",
    "    if unknown:\n",
    "        parser.print_usage()\n",
    "        print(f'Unrecognized arguments: {unknown[0]}')\n",
    "        exit(1)\n",
    "\n",
    "    utils.ContextScope._verbose = args.verbose\n",
    "    args = [args.__dict__[name] for name in flag_names]\n",
    "    args_string = \", \".join([f\"{name}: {value if value else '[None]'}\" for name, value in zip(flag_names, args)])\n",
    "    with utils.ContextScope(f\"Running {command_name}\",\n",
    "                            description=\"Args= \" + args_string,\n",
    "                            error=\"\"):\n",
    "        dependencies.check_dependencies()\n",
    "        endpoint(*args)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def handle_endpoints(command_name, flag_names, endpoint, level):\n",
    "    parser = argparse.ArgumentParser(prog=command_name)\n",
    "    for name in flag_names:\n",
    "        add_default_flag(parser, name)\n",
    "\n",
    "    add_default_flag(parser, \"verbose\")\n",
    "\n",
    "    args, unknown = parser.parse_known_args(sys.argv[level + 1:])\n",
    "    if unknown:\n",
    "        parser.print_usage()\n",
    "        print(f'Unrecognized arguments: {unknown[0]}')\n",
    "        exit(1)\n",
    "\n",
    "    utils.ContextScope._verbose = args.verbose\n",
    "    args = [args.__dict__[name] for name in flag_names]\n",
    "    args_string = \", \".join([f\"{name}: {value if value else '[None]'}\" for name, value in zip(flag_names, args)])\n",
    "    with utils.ContextScope(f\"Running {command_name}\",\n",
    "                            description=\"Args= \" + args_string,\n",
    "                            error=\"\"):\n",
    "        dependencies.check_dependencies()\n",
    "        endpoint(*args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "ename": "Exception",
     "evalue": "Unknown flag: C:\\Users\\nihyc\\Documents\\GitHub\\biof309\\genoml2\\outputs\\test_discrete_geno",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mException\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-49-cc78b167b26a>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mhandle_endpoints\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"genoml discrete supervised munge\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;34m\"C:\\\\Users\\\\nihyc\\\\Documents\\\\GitHub\\\\biof309\\\\genoml2\\\\outputs\\\\test_discrete_geno\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"impute\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"C:\\\\Users\\\\nihyc\\\\Documents\\\\GitHub\\\\biof309\\\\genoml2\\\\outputs\\\\examples/discrete/training\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"skip_prune\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"C:\\\\Users\\\\nihyc\\\\Documents\\\\GitHub\\\\biof309\\\\genoml2\\\\outputs\\\\examples/discrete/training_pheno.csv\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"addit\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"feature_selection\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"gwas\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"p\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"vif\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"iter\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m\"ref_cols_harmonize\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mfunctools\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpartial\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmunging\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmain\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdata_type\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"d\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m3\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m<ipython-input-46-ff749d63ed93>\u001b[0m in \u001b[0;36mhandle_endpoints\u001b[1;34m(command_name, flag_names, endpoint, level)\u001b[0m\n\u001b[0;32m    339\u001b[0m     \u001b[0mparser\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0margparse\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mArgumentParser\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mprog\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mcommand_name\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    340\u001b[0m     \u001b[1;32mfor\u001b[0m \u001b[0mname\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mflag_names\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 341\u001b[1;33m         \u001b[0madd_default_flag\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mparser\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mname\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    342\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    343\u001b[0m     \u001b[0madd_default_flag\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mparser\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"verbose\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-46-ff749d63ed93>\u001b[0m in \u001b[0;36madd_default_flag\u001b[1;34m(parser, flag_name)\u001b[0m\n\u001b[0;32m    333\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    334\u001b[0m     \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 335\u001b[1;33m         \u001b[1;32mraise\u001b[0m \u001b[0mException\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mf\"Unknown flag: {flag_name}\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    336\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    337\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mException\u001b[0m: Unknown flag: C:\\Users\\nihyc\\Documents\\GitHub\\biof309\\genoml2\\outputs\\test_discrete_geno"
     ]
    }
   ],
   "source": [
    "handle_endpoints(\"genoml discrete supervised munge\", [\"C:\\\\Users\\\\nihyc\\\\Documents\\\\GitHub\\\\biof309\\\\genoml2\\\\outputs\\\\test_discrete_geno\", \"impute\", \"C:\\\\Users\\\\nihyc\\\\Documents\\\\GitHub\\\\biof309\\\\genoml2\\\\outputs\\\\examples/discrete/training\", \"skip_prune\", \"C:\\\\Users\\\\nihyc\\\\Documents\\\\GitHub\\\\biof309\\\\genoml2\\\\outputs\\\\examples/discrete/training_pheno.csv\", \"addit\", \"feature_selection\", \"gwas\", \"p\", \"vif\", \"iter\",\"ref_cols_harmonize\"],functools.partial(munging.main, data_type=\"d\"), 3)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
